aesd-char-driver/main.c

```diff
@@ -47,23 +57,140 @@ int aesd_release(struct inode *inode, struct file *filp)
 ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
                 loff_t *f_pos)
 {
-	ssize_t retval = 0;
+	ssize_t read_len = 0;
+	size_t byte_offset = 0;
+	int err;
+
+
 	PDEBUG("read %zu bytes with offset %lld",count,*f_pos);
 	/**
 	 * TODO: handle read
 	 */
-	return retval;
+
+	struct aesd_dev *device = filp->private_data;
+
+	// acquire the lock
+	if ((err = mutex_lock_interruptible(&device->mutex)) != 0)
+	{
+		return err;
+	}	
+
+	struct aesd_buffer_entry *lc_entry = aesd_circular_buffer_find_entry_offset_for_fpos(&device->c_buff, 
+											*f_pos, &byte_offset);
+
+	if (lc_entry != NULL)
+	{
+		if((count + byte_offset) <= lc_entry->size)
+		{
+
+			read_len = count;
+
+		}
+		else
+		{
+			read_len = lc_entry->size - byte_offset;
+		}
+	}
+	else
+	{
+
+		// release the lock
+		mutex_unlock(&device->mutex);		
+		return 0;
+	}
+
+	if (copy_to_user(buf, (lc_entry->buffptr + byte_offset), read_len) != 0)
+	{
+		return -EFAULT;
+	}
+
+	*f_pos += read_len;
+
+	// release lock
+	mutex_unlock(&device->mutex);	
+
+	return read_len;
+
+
 }
 
 ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
                 loff_t *f_pos)
 {
-	ssize_t retval = -ENOMEM;
+
 	PDEBUG("write %zu bytes with offset %lld",count,*f_pos);
 	/**
-	 * TODO: handle write
+	 * TODO: handle write 
 	 */
-	return retval;
+	struct aesd_dev *device = NULL;
+	int err;
+	
+	device = filp->private_data;
+
+	// acquire the lock
+	if ((err = mutex_lock_interruptible(&device->mutex)) != 0)
+	{
+		return err;
+	}	
+
+	if (!(device->tmp = kzalloc(sizeof(char) * count, GFP_KERNEL)))
+	{
+		return -ENOMEM;
+	}
+
+	if (copy_from_user(device->tmp, buf, count) != 0)
+	{
+		kfree(device->tmp);
+		return -EFAULT;
+	}
+
+	// Note: In init module the aesd_device is init to zero
+	// During the first entry of this method, device->entry.size = 0
+	
+	char *swap = device->entry.buffptr;
+	if (!(device->entry.buffptr = kzalloc((sizeof(char) * (device->entry.size + count)), GFP_KERNEL)))
+	{
+		// if kzalloc fails reinstate the previous ptr value from swap
+		device->entry.buffptr = swap;
+		kfree(device->tmp);
+		return -ENOMEM;
```

Comment:
@SundarKrishnakumar what would happen to your mutex in this code path or the 3 error paths above?